### PR TITLE
Issue/cif 755

### PR DESCRIPTION
**Pull request by deepkamal**
Pagination bar in product listing component
## Description

Default count of products to be listed on screen is set to 6, this can be changed through dialog configuration of product list component.
Pagination bar is equipped with First, Previous, Next and Last anchors and individual page jump anchors. 
Author can define total number of products to be listed on page using property `pageSize`

## Related Issue

CIF-755 

## Motivation and Context

Pagination feature required on product listing
## How Has This Been Tested?

Added a test case in ProductListImplTest to verify paginationFields

## Screenshots (if appropriate):

image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes and the overall coverage did not decrease.
- [x] All unit tests pass on CircleCi.
- [x] I ran all tests locally and they pass.

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 1]
```diff
@@ -14,14 +14,16 @@
 
 package com.adobe.cq.commerce.core.components.internal.models.v1.productlist;
 
-import java.util.ArrayList;
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: Why all this import re-org?

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 2]
```diff
@@ -58,6 +49,7 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(ProductListImpl.class);
 
     private static final boolean SHOW_TITLE_DEFAULT = true;
+    private static int PAGE_SIZE_DEFAULT = 6;
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: Default can be `final`.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 3]
```diff
@@ -79,17 +71,30 @@
     private boolean showTitle;
     private MagentoGraphqlClient magentoGraphqlClient;
 
+    private int navPageCursor = 1;
+    private int navPageSize = 6;
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: Don't set it to a hardcoded number, either use the default declared above or leave empty, will be assigned later anyway.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 4]
```diff
@@ -79,17 +71,30 @@
...
 
         // get product template page
         productPage = Utils.getProductPage(currentPage);
         if (productPage == null) {
             productPage = currentPage;
         }
 
+        //check if pageCursor available in queryString, set to 1 if not.
+        if (request.getParameter("page") == null) {
+            this.navPageCursor = 1;
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: This is not needed `navPageCursor` is already 1.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 5]
```diff
@@ -79,17 +71,30 @@
...
+        } else {
+            this.navPageCursor = Integer.parseInt(request.getParameter("page"));
+        }
+
+
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: Remove blank line.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 6]
```diff
@@ -118,18 +133,19 @@ public boolean showTitle() {
     public Collection<ProductListItem> getProducts() {
         Collection<ProductListItem> listItems = new ArrayList<>();
 
+
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: Remove blank line

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 7]
```diff
@@ -118,18 +133,19 @@ public boolean showTitle() {
...
                 for (ProductInterface product : products.getItems()) {
                     listItems.add(new ProductListItemImpl(
-                        product.getSku(),
-                        product.getUrlKey(),
-                        product.getName(),
-                        product.getPrice().getRegularPrice().getAmount().getValue(),
-                        product.getPrice().getRegularPrice().getAmount().getCurrency().toString(),
-                        product.getSmallImage().getUrl(),
-                        productPage));
+                            product.getSku(),
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: Fix indent, don't use tabs.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 8]
```diff
@@ -139,28 +155,33 @@ public boolean showTitle() {
     /* --- GraphQL queries --- */
     private ProductPricesQueryDefinition generatePriceQuery() {
         return q -> q
-            .regularPrice(rp -> rp
-                .amount(a -> a
-                    .currency()
-                    .value()));
+                .regularPrice(rp -> rp
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: Fix indent, don't use tabs.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 9]
```diff
@@ -139,28 +155,33 @@ public boolean showTitle() {
...
     }
 
     private ProductInterfaceQueryDefinition generateProductQuery() {
         return q -> q
-            .id()
-            .name()
-            .smallImage(i -> i.url())
-            .urlKey()
-            .price(generatePriceQuery());
+                .id()
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: Fix indent, don't use tabs.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 10]
```diff
@@ -139,28 +155,33 @@ public boolean showTitle() {
...
     private CategoryTreeQueryDefinition generateProductListQuery() {
-        return q -> q
-            .id()
-            .description()
-            .name()
-            .productCount()
-            .products(categoryProductsQuery -> categoryProductsQuery.items(generateProductQuery()).totalCount());
+
+        CategoryTreeQuery.ProductsArgumentsDefinition pArgs = q -> q
+                .currentPage(this.navPageCursor)
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: Fix indent, don't use tabs.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 11]
```diff
@@ -201,10 +222,25 @@ private Integer parseCategoryId() {
...
+
+    @Override
+    public int getCurrentPage() {
+        return this.navPageCursor;
+    }
+
+    public int[] getPageList() {
+        return this.navPages;
+    }
+
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: Remove blank line

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 12]
```diff
@@ -201,10 +222,25 @@ private Integer parseCategoryId() {
...
             categoryId = Integer.parseInt(this.request.getRequestPathInfo().getSelectorString());
-        } catch (NumberFormatException nef) {
+        } catch (NullPointerException | NumberFormatException nef) {
             LOGGER.warn("Could not parse category id from current page selectors.");
         }
         return categoryId;
     }
+
+    @Override
+    public int getTotalCount() {
```
Comment (mhaack) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: Move these three methods up closer to the `getTitle` methods to keep the file structured.

File: bundles/core/src/test/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImplTest.java [hunk 13]
```diff
@@ -14,42 +14,33 @@
 
 package com.adobe.cq.commerce.core.components.internal.models.v1.productlist;
 
-import java.lang.reflect.Type;
```
Comment (mhaack) [inline bundles/core/src/test/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImplTest.java]: Again, fix import order.

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/_cq_dialog/.content.xml [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    <content
+        jcr:primaryType="nt:unstructured"
+        sling:resourceType="granite/ui/components/coral/foundation/fixedcolumns"
+        margin="{Boolean}false">
+        <items jcr:primaryType="nt:unstructured">
+            <column
+                jcr:primaryType="nt:unstructured"
+                sling:resourceType="granite/ui/components/coral/foundation/container">
+                <items jcr:primaryType="nt:unstructured">
+                    <pageSize
```
Comment (mhaack) [inline ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/_cq_dialog/.content.xml]: Why was this added to cq_dialog? We already have cq_design_dialog, I think the property config should go there, no need for an additional dialog.

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js [hunk 15]
```diff
@@ -0,0 +1,27 @@
+let getCookie = function (cname) {
```
Comment (mhaack) [inline ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js]: Why is that needed? Do we store the page state in a cookie?

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/_cq_dialog/.content.xml (hunk 14 above)
Comment (deepkamal) [reply ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/_cq_dialog/.content.xml]: it was discussed in forum and concluded that we want to keep `pageSize` authorable , thus cq_dialog was added, should we add this back in design dialog?

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js (hunk 15 above)
Comment (deepkamal) [reply ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js]: earlier it was designed like that, but later we discussed and agreed to keep page movements tracked via CGI params in query String, these can be used later, or can be removed - you suggest we remove this?

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/css/productlist-component.css [hunk 16]
```diff
@@ -111,3 +111,62 @@
     text-align: center;
 }
 
+.category__pagination_root {
```
Comment (LSantha) [inline ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/css/productlist-component.css]: Are these styles taken from Venia or they are custom styles created by you? If they are custom, we prefer to put them in a dedicated file, separately from the styles generated from the Venia components. The filename could be: productlist-aem.css .

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js (hunk 15 above)
Comment (LSantha) [reply ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js]: I opt for removing it. We can add it later if needed.

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+let setCookie = function (cname, cvalue, exdays) {
+    let d = new Date();
+    d.setTime(d.getTime() + (exdays * 24 * 60 * 60 * 1000));
+    let expires = "expires=" + d.toUTCString();
+    document.cookie = cname + "=" + cvalue + ";" + expires + ";path=/";
+};
+
+let switchToPageNumber = function (pageNum) {
+
+    window.location = "?page=" + pageNum;
```
Comment (LSantha) [inline ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js]: Guys, I wonder if using links on the buttons would be a cleaner approach than this? Then hovering over the button you can see the page number where the button takes you. I know Venia doesn't have this, so this is just a question.

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js (hunk 15 above)
Comment (mhaack) [reply ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js]: +1 for removing it. This is not a requirement for the ticket. 

@deepkamal "earlier it was designed like that"? This is no earlier version of this? What are you referring to?

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/_cq_dialog/.content.xml (hunk 14 above)
Comment (mhaack) [reply ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/_cq_dialog/.content.xml]: Yes `pageSize` should be authorable but having two dialogs for very similar author settings does not make any sense. In the design dialog it is authorable as well. I have no strong preference here, as long as we only have one dialog.

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js (hunk 15 above)
Comment (deepkamal) [reply ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js]: I'll remove it.
Pardon my English , by "earlier it was designed like that" I mean "earlier I started with using cookies, later switched to query string".

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java [hunk 18]
```diff
@@ -79,17 +71,30 @@
...
         productPage = Utils.getProductPage(currentPage);
         if (productPage == null) {
             productPage = currentPage;
         }
 
+        //check if pageCursor available in queryString, set to 1 if not.
+        if (request.getParameter("page") == null) {
+            this.navPageCursor = 1;
+        } else {
+            this.navPageCursor = Integer.parseInt(request.getParameter("page"));
```
Comment (herzog31) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: I would add a basic sanity check here. For example, the current page shouldn't be negative.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java (hunk 1 above)
Comment (deepkamal) [reply bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: this is done by IDE optimize import, as discussed, we will use IDE default in sequencing import classes

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/_cq_dialog/.content.xml (hunk 14 above)
Comment (deepkamal) [reply ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/_cq_dialog/.content.xml]: `pageSize` is only in cq_dialog, its not included in cq_design_dialog.
as discussed we are keeping in cq_dialog so it is authorable at page level.

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js (hunk 15 above)
Comment (deepkamal) [reply ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js]: nvhelper.js removed, page switching arithmetic now included in Model methods

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js (hunk 17 above)
Comment (deepkamal) [reply ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/js/navhelper.js]: Links are now generated without JS functions,

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java (hunk 18 above)
Comment (deepkamal) [reply bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImpl.java]: applied check #97182ad

File: ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/css/productlist-component.css (hunk 16 above)
Comment (deepkamal) [reply ui.apps/src/main/content/jcr_root/apps/venia/components/commerce/productlist/v1/productlist/clientlibs/dist/css/productlist-component.css]: the style classes are not auto-generated, a new CSS file created and added by name `productlist-aem.css` in dist directory, all styles moved in the file.

**Comment by herzog31** [comment]
Are you done with adding tests? If you check the coverage report, coverage is dropping by 2.4%. If you check https://codecov.io/gh/adobe/aem-core-cif-components/pull/35/diff?src=pr&el=tree#diff-YnVuZGxlcy9jb3JlL3NyYy9tYWluL2phdmEvY29tL2Fkb2JlL2NxL2NvbW1lcmNlL2NvcmUvY29tcG9uZW50cy9pbnRlcm5hbC9tb2RlbHMvdjEvcHJvZHVjdGxpc3QvUHJvZHVjdExpc3RJbXBsLmphdmE=, most of your pagination logic is not covered.

**Comment by deepkamal** [comment]
all requested changes are pushed, test coverage also increased. please review again @mhaack  @herzog31

File: bundles/core/src/test/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImplTest.java [hunk 19]
```diff
@@ -119,47 +110,51 @@ public void getProducts() {
     }
 
     @Test
-    public void testPaginationFields() {
+    public void testPagination() {
+
```
Comment (dplaton) [inline bundles/core/src/test/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImplTest.java]: Please remove the blank lines. Also, I don't think we need the comment anymore.

Comment (deepkamal) [reply bundles/core/src/test/java/com/adobe/cq/commerce/core/components/internal/models/v1/productlist/ProductListImplTest.java]: agreed, removed spaces and moved some lines of code out of setupPagination() method to initModel()
